webapi/src/main/scala/org/knora/webapi/messages/admin/responder/usersmessages/UsersMessagesADM.scala

```diff
@@ -555,6 +555,42 @@ case class UserADM(id: IRI,
         }
     }
 
+    /**
+      * Given an identifier, returns true if it is the same user, and false if not.
+      */
+    def isSelf(identifier: UserIdentifierADM): Boolean = {
+
+        val iriEquals = if (identifier.toIriOption.isDefined) {
```

Comment:
I was going to say that there's a shorter way to write this method:

```scala
val iriEquals = identifier.toIriOption.contains(id)
val emailEquals = identifier.toEmailOption.contains(email)
val usernameEquals = identifier.toUsernameOption.contains(id)

iriEquals || emailEquals || usernameEquals
```

But after thinking about it some more, I think there's a small design problem. Suppose you have two instances of `UserIdentifierADM` that represent the same user, but one instance contains the email address and the other instance contains the username. Then `isSelf` will return `false`, though it should return `true`. So I think the `isSelf` method belongs somewhere else, not in this class, so it can determine whether two `UserIdentifierADM` instances really represent the same user.

Also, I think it's not ideal that `value` is just a string that has to be parsed every time you use it (in `hasType`, `toIri`, and `toIriOption`). I think it would be simpler, safer, and more efficient to use types.

```scala
sealed trait UserIdentifierADM
case class UserIriADM(iri: IRI) extends UserIdentifierADM
case class UserEmailADM(email: String) extends UserIdentifierADM
case class UsernameADM(username: String) extends UserIdentifierADM
```

Looking at the places where a `UserIdentifierADM` is constructed, it looks like you almost always know what type of identifier you have, so you don't need `UserIdentifierADM` to figure it out for you. The exceptions that I see are in `UsersRouteADM` and `AuthenticationRouteV2`, where you get an identifier from the client, but you don't know what type it is. Here I would really suggest changing the route so that the client has to specify the type of the identifier. I think it doesn't seem very safe to rely on guessing the type of the identifier from its format.

Reply:
The idea behind `UserIdentifierADM` was exactly to allow the user to call a route and supply either an IRI, username or email, without the need to provide further information, as each uniquely identifies the user.

Reply:
> But after thinking about it some more, I think there's a small design problem. Suppose you have two instances of `UserIdentifierADM` that represent the same user, but one instance contains the email address and the other instance contains the username. Then `isSelf` will return `false`, though it should return `true`.

Why would it return `false`? If in the first case the `email` matches, then the result will be `true`. If in the second case the `username` matches, then also the result will be `true`.

Reply:
> The idea behind `UserIdentifierADM` was exactly to allow the user to call a route and supply either an IRI, username or email, without the need to provide further information, as each uniquely identifies the user.

What if my username is `foo@example.org`?

And there could be more types of identifiers in the future, creating more ambiguity.

> Why would it return `false`?

Sorry, I got confused, I thought the `isSelf` method was in `UserIdentifierADM`. Now I see it's in `UserADM`. But, it does seem to have a bug. I added a test to show this.

I think that if you used different classes for different types of identifiers, this kind of bug would be less likely.

Reply:
> But, it does seem to have a bug. I added a test to show this.

I've added already one in the commit before ;-)

Reply:
> I've added already one in the commit before ;-)

But my test still fails.

Reply:
> And there could be more types of identifiers in the future, creating more ambiguity.

hmm, then I should change the routes to something like this:

- `/admin/users/iri/<userIri>`
- `/admin/users/email/<email>`
- `/admin/users/username/<username>`
- `/admin/users/isni/<isni>`

and also the same structure for projects. There we have the same problem.

This will certainly make my life easier but will make it a bit harder for the consumers of the API.

Reply:
> hmm, then I should change the routes to something like this:

That seems OK to me. You could also do it with one route:

- `/admin/users?email=user%40example.org`
- `/admin/users?username=foo`

For logging in, you get a `POST` request, so the client can specify `email` , `username`, etc., instead of `identifier`.

Reply:
I think it's usually not a bad thing to make the client say explicitly what they want. It forces them to think about it more, which could prevent client bugs.